src/async-wrap.cc

```diff
@@ -201,42 +203,49 @@ Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
     Local<Value> enter_v = domain->Get(env()->enter_string());
     if (enter_v->IsFunction()) {
       enter_v.As<Function>()->Call(domain, 0, nullptr);
-      if (try_catch.HasCaught())
+      if (try_catch.HasCaught()) {
+        FatalException(env()->isolate(), try_catch);
         return Undefined(env()->isolate());
+      }
     }
   }
 
   if (ran_init_callback() && !pre_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
     pre_fn->Call(context, 0, nullptr);
     if (try_catch.HasCaught())
       FatalError("node::AsyncWrap::MakeCallback", "pre hook threw");
-    try_catch.SetVerbose(true);
   }
 
   Local<Value> ret = cb->Call(context, argc, argv);
 
+  // Allow domain and asyncwrap callbacks to run before throwing.
   if (try_catch.HasCaught()) {
-    return Undefined(env()->isolate());
+    cb_threw = true;
+    try_catch.Reset();
   }
 
   if (ran_init_callback() && !post_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
     post_fn->Call(context, 0, nullptr);
     if (try_catch.HasCaught())
       FatalError("node::AsyncWrap::MakeCallback", "post hook threw");
-    try_catch.SetVerbose(true);
   }
 
   if (has_domain) {
     Local<Value> exit_v = domain->Get(env()->exit_string());
     if (exit_v->IsFunction()) {
       exit_v.As<Function>()->Call(domain, 0, nullptr);
-      if (try_catch.HasCaught())
+      try_catch.Reset();
+      if (try_catch.HasCaught()) {
```

Comment:
If I understand correctly, `try_catch.HasCaught()` will return `false` after a call to `TryCatch::Reset`, so I'm not sure I understand what the use case is.